tensorcircuit/templates/lattice.py

```diff
@@ -587,151 +719,232 @@ def __init__(
-        for each site, populating the internal lattice data structures.
         """
-        current_index = 0
+        Generates all site information for the periodic lattice in a vectorized manner.
+        """
+        logger.debug("[DEBUG-LATTICE] Starting _build_lattice for TILattice.")
+        ranges = [backend.arange(s) for s in self.size]
+
+        # Generate a grid of all integer unit cell coordinates.
+        grid = backend.meshgrid(*ranges, indexing="ij")
+        all_cell_coords = backend.reshape(
+            backend.stack(grid, axis=-1), (-1, self.dimensionality)
+        )
 
-        # Iterate over all unit cell coordinates elegantly using np.ndindex
-        for cell_coord in np.ndindex(self.size):
-            cell_coord_arr = np.array(cell_coord)
-            # R = n1*a1 + n2*a2 + ...
-            cell_vector = np.dot(cell_coord_arr, self.lattice_vectors)
+        all_cell_coords = backend.cast(all_cell_coords, self.lattice_vectors.dtype)
 
-            # Iterate over the basis sites within the unit cell
-            for basis_index in range(self.num_basis):
-                basis_vec = self.basis_coords[basis_index]
+        cell_vectors = backend.tensordot(
+            all_cell_coords, self.lattice_vectors, axes=[[1], [0]]
+        )
+
+        cell_vectors = backend.cast(cell_vectors, self.basis_coords.dtype)
+
+        # Combine cell vectors with basis coordinates to get all site positions
+        # via broadcasting: (num_cells, 1, D) + (1, num_basis, D) -> (num_cells, num_basis, D)
+        all_coords = backend.expand_dims(cell_vectors, 1) + backend.expand_dims(
+            self.basis_coords, 0
+        )
 
-                # Calculate the real-space coordinate
-                coord = cell_vector + basis_vec
-                # Create a structured identifier
-                identifier = cell_coord + (basis_index,)
+        self._coordinates = backend.reshape(all_coords, (-1, self.dimensionality))
 
-                # Store site information
+        self._indices = []
+        self._identifiers = []
+        self._ident_to_idx = {}
+        current_index = 0
+
+        # Generate integer indices and tuple-based identifiers for all sites.
+        # e.g., identifier = (uc_x, uc_y, basis_idx)
+        size_ranges = [range(s) for s in self.size]
+        for cell_coord_tuple in itertools.product(*size_ranges):
+            for basis_index in range(self.num_basis):
+                identifier = cell_coord_tuple + (basis_index,)
                 self._indices.append(current_index)
                 self._identifiers.append(identifier)
-                self._coordinates.append(coord)
                 self._ident_to_idx[identifier] = current_index
                 current_index += 1
 
+        logger.debug(
+            f"[DEBUG-LATTICE] Finished _build_lattice. Total sites: {self.num_sites}"
+        )
+
     def _get_distance_matrix_with_mic(self) -> Coordinates:
```

Comment:
this function is not used as only the vectorized version is called? what about the performance difference between the two methods